From 3bede0edaa42fcf8fd15dce9d33db2ac8dca4a30 Mon Sep 17 00:00:00 2001
From: Alberto Milone <alberto.milone@canonical.com>
Date: Tue, 16 Jul 2019 16:42:17 +0200
Subject: [PATCH 1/1] Use kmem_cache_create_usercopy on 4.16+

Original author: Andreas Beckmann <anbe@debian.org>
Description: fixes "Bad or missing usercopy whitelist? Kernel memory exposure attempt
 detected from SLUB object 'nvidia_stack_cache'" on linux-image-4.16.0-2-*
 or newer that have disabled CONFIG_HARDENED_USERCOPY_FALLBACK
Bug-Debian: #901919, #899998
Bug: https://devtalk.nvidia.com/default/topic/1031067
---
 nv-linux.h | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/nv-linux.h b/nv-linux.h
index 4043bf1..99bc611 100644
--- a/nv-linux.h
+++ b/nv-linux.h
@@ -757,11 +757,18 @@ extern nv_spinlock_t km_lock;
                         0, 0, NULL, NULL);                      \
     }
 #elif (NV_KMEM_CACHE_CREATE_ARGUMENT_COUNT == 5)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 16, 0)
+#define NV_KMEM_CACHE_CREATE(kmem_cache, name, type)            \
+    {                                                           \
+        kmem_cache = kmem_cache_create_usercopy(name, sizeof(type), 0, 0, 0, sizeof(type), NULL); \
+    }
+#else
 #define NV_KMEM_CACHE_CREATE(kmem_cache, name, type)            \
     {                                                           \
         kmem_cache = kmem_cache_create(name, sizeof(type),      \
                         0, 0, NULL);                            \
     }
+#endif
 #else
 #error "NV_KMEM_CACHE_CREATE_ARGUMENT_COUNT value unrecognized!"
 #endif
-- 
2.20.1

